pkg/yang/types_builtin.go

```diff
@@ -506,32 +545,98 @@ func ParseNumber(s string) (n Number, err error) {
 		return n, errors.New("sign with no value")
 	}
 
+	n.Kind = Positive
+	ns := s
 	switch s[0] {
 	case '+':
-		s = s[1:]
+		ns = s[1:]
 	case '-':
 		n.Kind = Negative
-		s = s[1:]
+		ns = s[1:]
+	}
+	n.Value, err = strconv.ParseUint(ns, 0, 64)
+	if err == nil {
+		return n, nil
 	}
-	n.Value, err = strconv.ParseUint(s, 0, 64)
+
+	n.DecimalValue, n.FractionDigits, err = DecimalValueFromString(s, -1)
 	return n, err
 }
 
+// DecimalValueFromString returns an int64 representation of numStr and uint8
+// representation of the number of its fractional digits.
+// If fracDigRequired is >= 0, the number is represented with fracDigRequired
+// fractional digits, regardless of the precision of numStr, otherwise the
+// precision of numStr is used to set the number of fractional digits.
+// numStr must conform to Section 9.3.
+func DecimalValueFromString(numStr string, fracDigRequired int) (n *int64, fracDig *uint8, err error) {
+	s := numStr
+	decCount := strings.Count(s, ".")
+	var fd int
+	switch decCount {
+	case 0:
+	case 1:
+		fd = len(s) - 1 - strings.Index(s, ".")
+		if fracDigRequired < 0 {
+			fracDigRequired = fd
+		}
+		if fd > fracDigRequired {
+			return nil, nil, fmt.Errorf("%s has too much precision, expect <= %d fractional digits", s, fracDigRequired)
+		}
+	default:
+		return nil, nil, fmt.Errorf("%s is not a valid decimal number", s)
+	}
+
+	for i := 0; i < fracDigRequired-fd; i++ {
+		s += "0"
+	}
+
+	s = strings.Replace(s, ".", "", 1)
+
+	v, err := strconv.ParseInt(s, 0, 64)
+	if err != nil {
+		return nil, nil, fmt.Errorf("%s is not a valid decimal number: %s", numStr, err)
+	}
+
+	return Int64(v), Uint8(uint8(fd)), nil
+}
+
 // String returns n as a string in decimal.
 func (n Number) String() string {
+	out := ""
 	switch n.Kind {
 	case Negative:
-		return "-" + strconv.FormatUint(n.Value, 10)
+		out += "-"
 	case MinNumber:
 		return "min"
 	case MaxNumber:
 		return "max"
 	}
-	return strconv.FormatUint(n.Value, 10)
+	if n.IsDecimal() {
+		t := strconv.FormatInt(*n.DecimalValue, 10)
+		fd := int(*n.FractionDigits)
+		if fd > 0 {
+			return t[:len(t)-fd] + "." + t[len(t)-fd:]
+		}
+		return t
+	}
+	return out + strconv.FormatUint(n.Value, 10)
 }
 
-// Int returns n as an int64.  It returns an error if n overflows an int64.
+// DebugString returns n's internal represenatation as a string.
+func (n Number) DebugString() string {
```

Comment:
Move this into the _test file since it is only needed for tests.  There is no reason I can think of that we want to generally export this.